### PR TITLE
T24004 Comprehensive unit tests

### PR DESCRIPTION
**Pull request by pwithnall**
https://phabricator.endlessm.com/T24004

### CODE REVIEW

File: libeos-parental-controls/tests/app-filter.c
Comment (GeorgesStavracas) [inline libeos-parental-controls/tests/app-filter.c]: This discards `const`

File: libeos-parental-controls/tests/app-filter.c
Comment (GeorgesStavracas) [inline libeos-parental-controls/tests/app-filter.c]: This too.

File: libeos-parental-controls/tests/app-filter.c
Comment (GeorgesStavracas) [inline libeos-parental-controls/tests/app-filter.c]: And this

File: libeos-parental-controls/tests/app-filter.c
Comment (GeorgesStavracas) [inline libeos-parental-controls/tests/app-filter.c]: ... here too

**Review by GeorgesStavracas — APPROVED** [review]
These tests look good (very good, in fact), but there are 4 small compiler warnings being triggered.

**Comment by pwithnall** [comment]
> there are 4 small compiler warnings being triggered.

What warning flags are you using? I don’t get those warnings locally, and Jenkins doesn’t get them either. It would be good to enable such a warning in the top-level `meson.build`.

**Comment by GeorgesStavracas** [comment]
`-Wdiscarded-qualifiers` I guess. Notice that Meson outputs:

`Compiler for C supports arguments -Wno-discarded-qualifiers -Wdiscarded-qualifiers: YES`

So we're checking for this argument, but not using it.

Edit: and "by not using it", I mean that this specific argument is not present at the Meson-generated `compile_commands.json` file.

File: libeos-parental-controls/tests/app-filter.c
Comment (wjt) [inline libeos-parental-controls/tests/app-filter.c:304]: As an aside, I tend to do this kind of thing with one call to `g_test_add_data_func` per test data, perhaps because I am spoiled by [`@pytest.mark.parametrize`](https://docs.pytest.org/en/latest/parametrize.html) / [JUnit's less-pleasing equivalent](https://github.com/junit-team/junit4/wiki/Parameterized-tests). The nice thing about that approach is (of course) that if one case fails, the rest still run; it's also easier to see from the test name (either an integer index, or an explicit name on each test) which one is failing. If one of these tests fails in CI, the test name + assertion message will give no hint as to which it is.

File: libeos-parental-controls/tests/app-filter.c
Comment (wjt) [inline libeos-parental-controls/tests/app-filter.c]: This and the following line seem like a weird way to spell `g_variant_new ("(o)", object path)`

File: libeos-parental-controls/tests/app-filter.c
Comment (wjt) [inline libeos-parental-controls/tests/app-filter.c:501]: Using this API turns out pretty nicely. (It really does remind me of the Telepathy testing framework, [eg](https://github.com/TelepathyIM/telepathy-gabble/blob/master/tests/twisted/text/receipts.py#L30). Probably not a coincidence.)

File: libeos-parental-controls/app-filter.c
Comment (wjt) [inline libeos-parental-controls/app-filter.c:306]: If I compile the following test program with `-Wall`:

```c
char *foo (char *x) { return x; }

void  main(void)
{
  char *bar = foo(bar);
}
```

I get a warning:

```
bar.c: In function ‘main’:
bar.c:5:15: warning: ‘bar’ is used uninitialized in this function [-Wuninitialized]
   char *bar = foo(bar);
               ^~~~~~~~
```

More zealous warning level needed?

File: libeos-parental-controls/tests/app-filter.c
Comment (wjt) [inline libeos-parental-controls/tests/app-filter.c:50]: Nice, I've definitely open-coded this in at least 2 projects before.

**Comment by pwithnall** [comment]
Fixup pushed, thanks. The fixup will need to be squashed before this can be merged. This branch also still depends on https://gitlab.gnome.org/pwithnall/libglib-testing/merge_requests/4 being reviewed and merged in libglib-testing upstream.

File: libeos-parental-controls/tests/app-filter.c
Comment (wjt) [inline libeos-parental-controls/tests/app-filter.c:1009]: I feel like making `i` have the same type as `data->error_index` would be neater than needing this cast. (Do you even need this cast to a wider type, actually?)

I know that conceptually, an array index should be of type `size_t`, but in this case, we know its largest value is `4`.

File: libeos-parental-controls/tests/app-filter.c
Comment (wjt) [inline libeos-parental-controls/tests/app-filter.c:1091]: What's the rationale for doing some of these tests in the "worker thread" style, and some in this "async + inline replies" style? I guess it does demonstrate both models of GtDBusQueue usage.

File: libeos-parental-controls/tests/app-filter.c
Comment (pwithnall) [reply libeos-parental-controls/tests/app-filter.c:304]: I would like to do that, but the downsides are:
 1. It moves the test data away from the test code.
 2. The combination of `struct`s not being valid inline values and declarations needing to be at the start of the function means that `main()` becomes an incomprehensible mess.

On balance, I think those downsides are reason to use the approach I’ve used here, but I completely acknowledge the cons you give, and wish there was some way around them.

Maybe a `g_test_start_subtest()` call which could be used inside a test, a bit like `g_test_skip()`? That would fix your second con, but can’t fix the first. I’ve kind of done that already with the `g_test_message()` call in the loop, though.

Grumble, nothing is satisfying. :-(

File: libeos-parental-controls/tests/app-filter.c
Comment (pwithnall) [reply libeos-parental-controls/tests/app-filter.c]: You are very right.

Comment (pwithnall) [reply libeos-parental-controls/tests/app-filter.c]: Fixup pushed.

File: libeos-parental-controls/tests/app-filter.c
Comment (pwithnall) [reply libeos-parental-controls/tests/app-filter.c:1009]: Doing it this way means that the compiler forces you to handle the cases where `data->error_index < 0`, which is good, because we needed to handle them explicitly (i.e. they are possible and callers do set `error_index` to negative values).

File: libeos-parental-controls/tests/app-filter.c
Comment (pwithnall) [reply libeos-parental-controls/tests/app-filter.c:1091]: The worker thread style is needed whenever a test uses the synchronous libeos-parental-controls API. Any other test which can sensibly reuse the same worker thread function also uses it. Everything else is done in the ‘inline replies’ style because it keeps the client/server conversation flow in a single function, so it should be a bit more obvious what’s happening.

And yes, it is partially to demonstrate both models of `GtDBusQueue` usage.

Does that make sense?

File: libeos-parental-controls/tests/app-filter.c
Comment (wjt) [reply libeos-parental-controls/tests/app-filter.c:304]: Well, you could move the definition to the top level and leave it defined here, just reference it from main(). But fair enough.

File: libeos-parental-controls/tests/app-filter.c
Comment (wjt) [reply libeos-parental-controls/tests/app-filter.c:1009]: Fair enough.

File: libeos-parental-controls/tests/app-filter.c
Comment (wjt) [reply libeos-parental-controls/tests/app-filter.c:1091]: Yep!

**Comment by GeorgesStavracas** [comment]
Hm, this branch breaks the GNOME Settings panel. The app filter is being denied:

`(gnome-control-center:20425): user-accounts-cc-panel-WARNING **: Error retrieving app filter for user 'georges': Not allowed to query app filter data for user 1001`

**Comment by pwithnall** [comment]
> Hm, this branch breaks the GNOME Settings panel. The app filter is being denied:
> 
> `(gnome-control-center:20425): user-accounts-cc-panel-WARNING **: Error retrieving app filter for user 'georges': Not allowed to query app filter data for user 1001`

Have you installed the updated `accounts-service/com.endlessm.ParentalControls.AppFilter.xml` file from eos-parental-controls and restarted `accounts-daemon.service` since then? This is likely being caused by the renaming of the properties — if accountsservice is returning the old property names to libeos-parental-controls, if will assume the lack of new properties means that permission was denied to access them.

File: libeos-parental-controls/app-filter.c
Comment (pwithnall) [reply libeos-parental-controls/app-filter.c:306]: Hmm, interesting. Putting that code into `app-filter.c` gives me a warning. Undoing the `old_flatpak_app` fix doesn’t give me a warning. Putting the following code (i.e. wrapping your code in a loop) into `app-filter.c` also *doesn’t* give me a warning:
```
  for (gsize i = 0; i < 10; i++)
    {
      char *bar = g_strstrip (bar);
    }
```

So it looks like the compiler’s heuristics for detecting `-Wmaybe-uninitialized` get broken by loops.

I’m using GCC 8.2.1. The bug might be https://gcc.gnu.org/bugzilla/show_bug.cgi?id=54554, since I’m compiling at `-O0`; but looking at the list of ‘uninitialized’ bugs in the GCC Bugzilla, there’s enough there that it’s either been reported already, or reporting it would not get it fixed.

In any case, Meson does seem to be successfully passing the warning flags in to the compiler command.

**Comment by GeorgesStavracas** [comment]
> Have you installed the updated accounts-service/com.endlessm.ParentalControls.AppFilter.xml file from eos-parental-controls and restarted accounts-daemon.service since then? This is likely being caused by the renaming of the properties — if accountsservice is returning the old property names to libeos-parental-controls, if will assume the lack of new properties means that permission was denied to access them.

That, and a corrupted .xml file (missing the last couple lines) were breaking everything. Tested on a fresh install, with a correct file, restarted the accountsservice daemon, and everything's working as it should.

**Comment by GeorgesStavracas** [comment]
Please squash the fixups so it can be merged

**Comment by pwithnall** [comment]
Squashed and rebased on master. I’ll wait for the tests to pass then merge. Thanks.

**Pull request merged**
